src/formatter/simple.rs

```diff
@@ -1,361 +1,428 @@
+use std::io;
+use std::mem;
+
+use colored::*;
+
 use events::{Event, EventHandler};
 use formatter::formatter::Formatter;
-use runner;
-use std::io;
+use context_report::ContextReport;
+use example_report::ExampleReport;
+
+use suite::SuiteInfo;
+use context::ContextInfo;
+use example::ExampleInfo;
+
+#[derive(Clone, Debug)]
+enum ScopeInfo {
+    Suite(SuiteInfo),
+    Context(ContextInfo),
+    Example(ExampleInfo),
+}
 
 pub struct Simple<'a, Io: io::Write + 'a> {
     buf: &'a mut Io,
-    pub name_stack: Vec<String>,
-    pub failures: Vec<String>,
+    name_stack: Vec<ScopeInfo>,
+    failed: Vec<Vec<ScopeInfo>>,
 }
 
 impl<'a, T: io::Write> Simple<'a, T> {
     pub fn new(buf: &mut T) -> Simple<T> {
         Simple {
             buf: buf,
             name_stack: vec![],
-            failures: vec![],
+            failed: vec![],
         }
     }
 
-    fn failures_summary(&self) -> String {
-        let res = String::with_capacity(100);
-        let mut idx = 0;
-        self.failures
-            .iter()
-            .map(|fail| {
-                idx += 1;
-                format!("  {}) {}\n", idx, fail)
-            })
-            .fold(res, |mut acc, elt| {
-                acc.push_str(&elt);
-                acc
-            })
+    fn padding(depth: usize) -> String {
+        "  ".repeat(depth)
     }
 
-    fn write_summary(&mut self, result: runner::RunnerResult) -> String {
-        let (res, report) = match result {
-            Ok(report) => ("ok", report),
-            Err(report) => ("FAILED", report),
-        };
-
-        format!("\n\ntest result: {}. {} examples; {} passed; {} failed;",
-                res,
-                report.total_tests,
-                report.success_count,
-                report.error_count)
+    fn enter_suite(&mut self, info: &SuiteInfo) {
+        self.name_stack.push(ScopeInfo::Suite(info.clone()));
+        let _ = writeln!(self.buf, "\nrunning tests");
+        let label: &str = info.label.into();
+        let _ = writeln!(self.buf, "{} {:?}:", label, info.name);
     }
-}
 
-impl<'a, T: io::Write> EventHandler for Simple<'a, T> {
-    fn trigger(&mut self, event: &Event) {
-        // FIXME: do something with the io::Error ?
-        let _ = match *event {
-            Event::StartRunner => writeln!(self.buf, "Running tests:\n"),
-            Event::StartDescribe(ref name) |
-            Event::StartTest(ref name) => {
-                self.name_stack.push(name.clone());
-                Ok(())
-            }
-            Event::EndTest(result) => {
-                if result.is_err() && !self.name_stack.is_empty() {
-                    let failure_name = self.name_stack.join(" | ");
-                    self.failures.push(failure_name);
+    fn exit_suite(&mut self, report: &ContextReport) {
```

Comment:
I have a second PR almost ready however which completely parallelizes the entire test suite execution (configurable) via *rayon*. It completely messes up the console log for obvious reasons. With this in mind I'd suggest to just ignore this one for now when deciding when and how to merge this PR, as I plan to rework this part (and add unit tests) in the very next PR to follow. But for the sake of clarity I'd rather not any further major refactors to this PR.

The next PR also turns the entire suite/context/example hierarchy immutable and makes the runner execution immutable (as in `…(&self, …)`, plus some `Arc<Mutex<…>>` on the `EventHandler`s) as well. This whole immutability thing is what enables the fearless concurrency _cough_ to begin with.